skcosmo/preprocessing/FlexibleScaler.py

```diff
@@ -0,0 +1,242 @@
+                + " instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator."
+            )
+        if (self.n_samples_seen_ != X_tr.shape[0]) or (
+            self.n_features_ != X_tr.shape[1]
+        ):
+            raise ValueError("X shape does not match training shape")
+        return X_tr * self.scale_ + self.mean_
+
+
+class KernelFlexibleCenterer(TransformerMixin, BaseEstimator):
+    """Kernel centering method, similar to KernelCenterer,
+    but with additional parameters, relative to which centering
+    is carried out:
+
+    :param ref_cmean: an array with means for each column.
+    :type ref_cmean: array of shape (insert shape here)
+
+    :param ref_rmean: an array with means for each row.
+    :type ref_rmean: array of shape (insert shape here)
+
+    :param ref_mean: an average for the whole kernel matrix
+    :type ref_mean: array
+    """
+
+    def __init__(self, *, ref_cmean=None, ref_rmean=None, ref_mean=None):
+        """Initialize KernelFlexibleCenterer."""
+        pass
+
+    def fit(self, K=None, y=None, ref_cmean=None, ref_rmean=None, ref_mean=None):
+        """Fit KernelFlexibleCenterer
+
+        :param K: Kernel matrix
+        :type K: ndarray of shape (n_samples, n_samples)
+        :param y: ignored
+        :param ref_cmean: an array with means for each column.
+        :type ref_cmean: array of shape (insert shape here)
+        :param ref_rmean: an array with means for each row.
+        :type ref_rmean: array of shape (insert shape here)
+        :param ref_mean: an average for the whole kernel matrix
+        :type ref_mean: array
+        :return: itself
+        """
+        if K is not None:
+            if K.shape[0] != K.shape[1]:
+                raise ValueError(
+                    "The reference kernel is not square, and does not define a RKHS"
+                )
+        else:
+            assert (
+                ref_cmean is not None and ref_rmean is not None and ref_mean is not None
+            )
+
+            if len(ref_rmean) != len(ref_cmean):
+                raise ValueError(
+                    "The reference kernel is not square, and does not define a RKHS"
+                )
+        self.ref_cmean = ref_cmean
+        self.ref_rmean = ref_rmean
+        self.ref_mean = ref_mean
+        self.reference_shape = []
+        if ref_rmean is None:
+            self.reference_shape.append(K.shape[0])
+        else:
+            self.reference_shape.append(len(ref_rmean))
+        if ref_cmean is None:
+            self.reference_shape.append(K.shape[1])
+        else:
+            self.reference_shape.append(len(ref_cmean))
+        self.reference_shape = tuple(self.reference_shape)
+        if self.ref_cmean is None:
+            self.ref_cmean = K.mean(axis=0)
+        if self.ref_rmean is None:
+            self.ref_rmean = K.mean(axis=1)
+        if self.ref_mean is None:
+            self.ref_mean = K.mean()
```

Comment:
So we need to define some priority in the arguments here. If both K and ref_cmean/ref_rmean/ref_mean are supplied, which one is used? Do we check the shapes of the two sets with respect to each other? I would suggest that we say ref_cmean/ref_rmean/ref_mean overrides K, which would lead to the code looking like:

```suggestion
        if K is not None:
            if K.shape[0] != K.shape[1]:
                raise ValueError(
                    "The reference kernel is not square, and does not define a RKHS"
                )
                
            if ref_cmean is not None:
                if K.shape[0] != len(ref_cmean):
                    raise ValueError("The supplied column mean does not match the supplied kernel.")
            else:
                ref_cmean = K.mean(axis=0)
                
            if ref_rmean is not None:
                if K.shape[1] != len(ref_rmean):
                    raise ValueError("The supplied row mean does not match the supplied kernel.")
            else:
                ref_rmean = K.mean(axis=1)
                
            if ref_mean is None:
                ref_mean = K.mean()
                
        else:
            assert (
                ref_cmean is not None and ref_rmean is not None and ref_mean is not None
            )

            if len(ref_rmean) != len(ref_cmean):
                raise ValueError(
                    "The reference kernel is not square, and does not define a RKHS"
                )
                
        self.reference_shape = [len(ref_rmean), len(ref_cmean)]
        self.ref_cmean = ref_cmean
        self.ref_rmean = ref_rmean
        self.ref_mean = ref_mean
        
```

Reply:
Yes, that sounds really nice.